Tools/build/check_warnings.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+"""
+Parses compiler output with -fdiagnostics-format=json and checks that warnings exist
+only in files that are expected to have warnings.
+"""
+import sys
+from pathlib import Path
+import argparse
+import json
+import re
+
+def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
+    """
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
+
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+
+    # Regex to find json arrays at the top level of the file in the compiler output
+    json_arrays = re.findall(r'\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]', compiler_output)
+    compiler_warnings = []
+    for array in json_arrays:
+        try:
+            json_data = json.loads(array)
+            json_objects_in_array = [entry for entry in json_data]
+            compiler_warnings.extend([entry for entry in json_objects_in_array if entry.get('kind') == 'warning'])
+        except json.JSONDecodeError:
+            continue  # Skip malformed JSON
+
+    return compiler_warnings
+
+def get_unexpected_warnings(
```

Comment:
I see a lot of overlap between `get_unexpected_warnings` and `get_unexpected_improvements`, should they be combined somehow or the warning gathering abstracted out?

Reply:
That is a good point. I refactored the code to extract the filename from the warnings in a separate function. This is only done once and passed to the two routines to check for regressions and improvements